shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Rupd.hs

```diff
@@ -94,25 +113,29 @@ rupdTransition ::
   TransitionRule (RUPD era)
 rupdTransition = do
   TRC (RupdEnv b es, ru, s) <- judgmentContext
-  (slotsPerEpoch, slot, maxLL) <- liftSTS $ do
+  (slotsPerEpoch, slot, maxLL, asc, lastblock) <- liftSTS $ do
     ei <- asks epochInfo
     sr <- asks randomnessStabilisationWindow
     e <- epochInfoEpoch ei s
     slotsPerEpoch <- epochInfoSize ei e
     slot <- epochInfoFirst ei e <&> (+* (Duration sr))
+    (_first, lastblock) <- lift (epochInfoRange ei e)
     maxLL <- asks maxLovelaceSupply
-    return (slotsPerEpoch, slot, maxLL)
-  if s <= slot
-    then pure ru
-    else case ru of
-      SNothing ->
-        SJust
-          <$> liftSTS
-            ( runProvM $
-                createRUpd
-                  slotsPerEpoch
-                  b
-                  es
-                  (Coin (fromIntegral maxLL))
-            )
-      SJust _ -> pure ru
+    asc <- asks activeSlotCoeff
+    return (slotsPerEpoch, slot, maxLL, asc, lastblock)
+  let maxsupply = Coin (fromIntegral maxLL)
+  case (s <= slot, s == lastblock) of
+    -- Waiting for the stabiliy point, do nothing, keep waiting
+    (True, _) -> pure SNothing
+    -- We are in the last block, finish everything up
+    (False, True) ->
+      case ru of
+        SNothing -> liftSTS $ runProvM $ SJust <$> completeStep (startStep slotsPerEpoch b es maxsupply asc)
```

Comment:
@nc6 I think I found a cleaner way of calculating the pulses per block, one that avoids using `k` (which is nice since then we don't have to pass an argument to `createRUpd` that isn't there in the spec).

`lastBlock` above should be named `lastSlot`. So `lastSlot - slot` is the last `6k/f` many _slots_ in the epoch, and so we expect `(lastSlot - slot) / f` many blocks in this period. We can divide the number of pools by the expected number of blocks.

Reply:
Ok this is done, I'd appreciate if you'd take a look to see that It is done correctly.